File: src/PristaneLaverieSmart.Domain/Entities/Booking.cs
using System.ComponentModel;
using PristaneLaverieSmart.Domain.Common;
using PristaneLaverieSmart.Domain.Enums;
using PristaneLaverieSmart.Domain.Events;

namespace PristaneLaverieSmart.Domain.Entities;

public class Booking: Entity
{
    public Guid Id {get; set;} = Guid.NewGuid();
    public Guid MachinedId {get; set;}
    public DateTimeOffset StartTime {get; set;}
    public DateTimeOffset EndTime {get; set;}
    public string CustomerName {get; set;} = string.Empty;
    public BookingStatus Status {get; set;} = BookingStatus.Received;

    public static Booking Create(Guid machineId, DateTimeOffset start, DateTimeOffset end, string customerName)
    {
        var booking = new Booking
        {
            MachinedId = machineId,
            StartTime = start,
            EndTime = end,
            CustomerName = customerName,
            Status = BookingStatus.Received
        };

        booking.AddDomainEvent(new BookingCreatedDomainEvent(booking.Id, booking.MachinedId, booking.StartTime, booking.EndTime));
        return booking;
    }

    
    public void Cancel()
    {
        Status = BookingStatus.Cancelled;
        AddDomainEvent(new BookingCancelledDomainEvent(Id, MachinedId));
    }
    public void Complete()
    {
        Status = BookingStatus.Completed;
        AddDomainEvent(new BookingCompletedDomainEvent(Id, MachinedId));
    }




}

File: domain/src/main/java/ch/pristane/laverie/smart/domain/entities/Booking.java
package ch.pristane.laverie.smart.domain.entities;

import ch.pristane.laverie.smart.domain.enums.BookingStatus;

import java.time.OffsetDateTime;
import java.util.UUID;

public class Booking {
    private UUID id;
    private UUID machineId;
    private OffsetDateTime startTime;
    private OffsetDateTime endTime;
    private String customerName;
    private BookingStatus status;

    public Booking() {
    }

    public Booking(UUID id, UUID machineId, OffsetDateTime startTime, OffsetDateTime endTime, String customerName, BookingStatus status) {
        this.id = id;
        this.machineId = machineId;
        this.startTime = startTime;
        this.endTime = endTime;
        this.customerName = customerName;
        this.status = status;
    }

    public static Booking create(UUID machineId, OffsetDateTime startTime, OffsetDateTime endTime, String customerName) {
        return new Booking(
                UUID.randomUUID(),
                machineId,
                startTime,
                endTime,
                customerName,
                BookingStatus.RECEIVED
        );
    }

    public void cancel() {
        if (status == BookingStatus.CANCELLED) {
            throw new IllegalStateException("Booking is already cancelled.");
        }
        if (status == BookingStatus.COMPLETED) {
            throw new IllegalStateException("Completed bookings cannot be cancelled.");
        }
        status = BookingStatus.CANCELLED;
    }

    public void complete() {
        if (status == BookingStatus.COMPLETED) {
            throw new IllegalStateException("Booking is already completed.");
        }
        if (status == BookingStatus.CANCELLED) {
            throw new IllegalStateException("Cancelled bookings cannot be completed.");
        }
        status = BookingStatus.COMPLETED;
    }

    public UUID getId() {
        return id;
    }

    public UUID getMachineId() {
        return machineId;
    }

    public OffsetDateTime getStartTime() {
        return startTime;
    }

    public OffsetDateTime getEndTime() {
        return endTime;
    }

    public String getCustomerName() {
        return customerName;
    }

    public BookingStatus getStatus() {
        return status;
    }
}